### PR TITLE
Add transparency support for LightmapGI

### LINKED ISSUE #77590: LightmapGI does not take transparency into account when baking shadows.
### Godot version

v4.0.3.stable - v4.3.dev3

### System information

Godot v4.3.dev3 - Fedora Linux 39 (KDE Plasma) - Wayland - Vulkan (Forward+) - dedicated AMD Radeon RX 580 Series (RADV POLARIS10) () - Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz (8 Threads)

### Issue description

As the title says, LightmapGI bakes shadows for objects with transparent materials as if they are 100% opaque.
I tried baking with every transparency mode possible, Alpha, Scissor, Hashed, Depth Prepass, none of it makes a difference.
This worked fine in Godot 3, so there is definitely a bug here.

Godot 3:
image

Godot 4:
image

### Steps to reproduce

1. Download the included MRP's, both Godot 3 and Godot 4 versions are provided.
2. Load up the projects and compare results.

### Minimal reproduction project

[mrps.zip](https://github.com/godotengine/godot/files/11591010/mrps.zip)

### PR DESCRIPTION
**Pull request by guerro323**
Fixes https://github.com/godotengine/godot/issues/77590

Scene from the original issue
|Current `master`|This PR `size=0`|This PR `size=0.1`|
:---:|:---:|:---:|
image | image | image  

### Test Scenes
image
image

___

The culling logic was updated in the Lightmapper to correctly support `Back` and `Disabled` cull modes.  
To preserve the current behavior and to not have any breaking change, the first direct light rays will retain the old behavior where light is blocked even if it should be culled.  
This behavior is still needed otherwise there would be holes in the shadows of penetrated meshes (which can be quite common) and would need the material to be updated to `Disabled` cull mode, which may not be intuitive in most cases.  

TL;DR: There **should** be no breaking changes on existing scenes. This do need some more testing but all my scenes looked correct.

___
*note: I've split the colored shadows into another PR (https://github.com/godotengine/godot/pull/90132)*

### CODE REVIEW

File: doc/classes/ProjectSettings.xml [hunk 1]
```diff
@@ -2513,6 +2513,9 @@
 		<member name="rendering/lightmapping/bake_performance/max_rays_per_probe_pass" type="int" setter="" getter="" default="64">
 			The maximum number of rays that can be thrown per pass when baking dynamic object lighting in [LightmapProbe]s with [LightmapGI]. Depending on the scene, adjusting this value may result in higher GPU utilization when baking lightmaps, leading to faster bake times.
 		</member>
+		<member name="rendering/lightmapping/bake_performance/max_transparency_rays" type="int" setter="" getter="" default="64">
+			The maximum number of retries rays that can be thrown per pass when hitting a transparent surface when baking lightmaps with [LightmapGI]. Depending on the scene, adjusting this value may result in higher GPU utilization when baking lightmaps, leading to faster bake times.
```
Comment (AThousandShips) [inline doc/classes/ProjectSettings.xml:2517]:
```suggestion
			The maximum number of retry rays that can be thrown per pass when hitting a transparent surface when baking lightmaps with [LightmapGI]. Depending on the scene, adjusting this value may result in higher GPU utilization when baking lightmaps, leading to faster bake times.
```
Only the main noun is plural, i.e. it's not "dogs catchers", but "dog catchers"

**Review by Calinou — COMMENTED** [review]
Tested locally, I can't get transparent shadows to show up for a static object regardless of its transparency mode. Shadows are always fully opaque. This is with Alpha, but the same happens with Alpha Scissor:

Screenshot_20240404_153953 webp

The DirectionalLight3D in the scene has the Static bake mode. I'm using the Compatibility rendering method in the test project (it bakes via a local RenderingDevice that uses Vulkan), but this also happens with Forward+. I've also tried disabling **Use Texture for Bounces** in LightmapGI (then baking lightmaps again) to no avail.

**Testing project:** [test_transparent_shadows.zip](https://github.com/godotengine/godot/files/14872112/test_transparent_shadows.zip)

**Comment by guerro323** [comment]
@Calinou I'm not able to repro it, even from the artifact builds.
|No changes Forward+|Increased the alpha range from the noise texture¹|Alpha scissor¹|
|:---:|:---:|:---:|
image | image | image

___
¹ If the shadows don't match it's because the bottom of the cube has a different view of the noise texture.

I noticed that alpha scissor didn't correctly work on the current commit (it only discarded alpha under the threshold but kept the same alpha if not, now it's binary) so it's fixed.
Also compatibility seems to not work (it just totally discard the alpha so it's completely transparent), it is using a custom RD specialized for it? I thought it would have used either the Forward+ or Mobile one.

**Comment by Calinou** [comment]
> Also compatibility seems to not work (it just totally discard the alpha so it's completely transparent), it is using a custom RD specialized for it? I thought it would have used either the Forward+ or Mobile one.

Baking lightmaps while using Compatibility was implemented in https://github.com/godotengine/godot/pull/87386.

**Comment by guerro323** [comment]
Baking on compatibility works now
|No changes Compatibility|Increased the alpha range from the noise texture¹|Alpha scissor¹|
|:---:|:---:|:---:|
image | image | image

___
¹ If the shadows don't match it's because the bottom of the cube has a different view of the noise texture.

**Comment by patwork** [comment]
> The culling logic was updated in the Lightmapper to correctly support Back and Disabled cull modes.

Will it fix https://github.com/godotengine/godot/issues/89402 ?

**Comment by lawnjelly** [comment]
I had limited time, but after the rendering meeting yesterday I had a look through and this looked fairly sensible from what I could see. It is also possible to do translucent colors in a similar way, but maybe for a later PR.

**Comment by jcostello** [comment]
> I had limited time, but after the rendering meeting yesterday I had a look through and this looked fairly sensible from what I could see. It is also possible to do translucent colors in a similar way, but maybe for a later PR.
> 
> I think @clayjohn mentioned wanting to double check some of the changes didn't affect some of the other GI paths.

There is a PR for translucent colors

File: modules/lightmapper_rd/lightmapper_rd.cpp [hunk 2]
```diff
@@ -458,7 +459,15 @@ void LightmapperRD::_create_acceleration_structures(RenderingDevice *rd, Size2i
 			t.max_bounds[0] = taabb.position.x + MAX(taabb.size.x, 0.0001);
 			t.max_bounds[1] = taabb.position.y + MAX(taabb.size.y, 0.0001);
 			t.max_bounds[2] = taabb.position.z + MAX(taabb.size.z, 0.0001);
-			t.pad0 = t.pad1 = 0; //make valgrind not complain
+
+			t.cull_mode = RendererRD::MaterialStorage::ShaderData::CULL_BACK;
+
+			RID material = mi.data.material[i];
+			if (material.is_valid()) {
+				RendererRD::MaterialStorage::ShaderData *shader_data = material_storage->material_get_shader_data(mi.data.material[i]);
```
Comment (clayjohn) [inline modules/lightmapper_rd/lightmapper_rd.cpp:467]: This crashes when baking from the compatibility renderer. Remember that that the compatibility renderer has its own storage classes. If we need to request this data from the renderer, then it needs to be exposed through the common ancestor MaterialStorage class

Backtrace:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.3.dev.custom_build (e34b9d1f31eada4e3966f1f4d2ac8bad9b8c6eed)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x731fd6e42520] (??:0)
[2] RID_Alloc<RendererRD::MaterialStorage::Material, true>::get_or_null(RID const&, bool) (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/atomic_base.h:212)
[3] RID_Owner<RendererRD::MaterialStorage::Material, true>::get_or_null(RID const&) (/home/clayjohn/dev/godot/./core/templates/rid_owner.h:451)
[4] RendererRD::MaterialStorage::get_material(RID) (/home/clayjohn/dev/godot/servers/rendering/renderer_rd/storage_rd/material_storage.h:251)
[5] RendererRD::MaterialStorage::material_get_shader_data(RID) (/home/clayjohn/dev/godot/servers/rendering/renderer_rd/storage_rd/material_storage.cpp:2130)
[6] LightmapperRD::_create_acceleration_structures(RenderingDevice*, Vector2i, int, AABB&, int, unsigned int, Vector<LightmapperRD::Probe>&, Lightmapper::GenerateProbes, Vector<int>&, Vector<int>&, RID&, RID&, RID&, RID&, RID&, RID&, RID&, RID&, RID&, bool (*)(float, String const&, void*, bool), void*) (/home/clayjohn/dev/godot/modules/lightmapper_rd/lightmapper_rd.cpp:467)
[7] LightmapperRD::bake(Lightmapper::BakeQuality, bool, float, int, float, float, int, bool, bool, Lightmapper::GenerateProbes, Ref<Image> const&, Basis const&, bool (*)(float, String const&, void*, bool), void*, float) (/home/clayjohn/dev/godot/modules/lightmapper_rd/lightmapper_rd.cpp:1189)
[8] LightmapGI::bake(Node*, String, bool (*)(float, String const&, void*, bool), void*) (/home/clayjohn/dev/godot/scene/3d/lightmap_gi.cpp:1070)
[9] LightmapGIEditorPlugin::_bake_select_file(String const&) (/home/clayjohn/dev/godot/editor/plugins/lightmap_gi_editor_plugin.cpp:71)
[10] LightmapGIEditorPlugin::_bake() (/home/clayjohn/dev/godot/editor/plugins/lightmap_gi_editor_plugin.cpp:120)
[11] void call_with_variant_args_helper<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (/home/clayjohn/dev/godot/./core/variant/binder_common.h:309)
[12] void call_with_variant_args_dv<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (/home/clayjohn/dev/godot/./core/variant/binder_common.h:452)
[13] MethodBindT<>::call(Object*, Variant const**, int, Callable::CallError&) const (/home/clayjohn/dev/godot/./core/object/method_bind.h:344)
[14] Object::callp(StringName const&, Variant const**, int, Callable::CallError&) (/home/clayjohn/dev/godot/core/object/object.cpp:837)
[15] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/clayjohn/dev/godot/core/variant/callable.cpp:69)
[16] Object::emit_signalp(StringName const&, Variant const**, int) (/home/clayjohn/dev/godot/core/object/object.cpp:1201)
[17] Node::emit_signalp(StringName const&, Variant const**, int) (/home/clayjohn/dev/godot/scene/main/node.cpp:3901)
[18] Error Object::emit_signal<>(StringName const&) (/home/clayjohn/dev/godot/./core/object/object.h:933)
[19] BaseButton::_pressed() (/home/clayjohn/dev/godot/scene/gui/base_button.cpp:139)
[20] BaseButton::on_action_event(Ref<InputEvent>) (??:?)
[21] BaseButton::gui_input(Ref<InputEvent> const&) (/home/clayjohn/dev/godot/scene/gui/base_button.cpp:69)
[22] Control::_call_gui_input(Ref<InputEvent> const&) (/home/clayjohn/dev/godot/scene/gui/control.cpp:1796)
[23] Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (/home/clayjohn/dev/godot/scene/main/viewport.cpp:1538)
[24] Viewport::_gui_input_event(Ref<InputEvent>) (/home/clayjohn/dev/godot/scene/main/viewport.cpp:1805)
[25] Viewport::push_input(Ref<InputEvent> const&, bool) (/home/clayjohn/dev/godot/scene/main/viewport.cpp:3229)
[26] Window::_window_input(Ref<InputEvent> const&) (/home/clayjohn/dev/godot/scene/main/window.cpp:1631)
[27] void call_with_variant_args_helper<Window, Ref<InputEvent> const&, 0ul>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/home/clayjohn/dev/godot/./core/variant/binder_common.h:304)
[28] void call_with_variant_args<Window, Ref<InputEvent> const&>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, int, Callable::CallError&) (/home/clayjohn/dev/godot/./core/variant/binder_common.h:419)
[29] CallableCustomMethodPointer<Window, Ref<InputEvent> const&>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/clayjohn/dev/godot/./core/object/callable_method_pointer.h:104)
[30] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/clayjohn/dev/godot/core/variant/callable.cpp:58)
[31] Variant Callable::call<Ref<InputEvent> >(Ref<InputEvent>) const (/home/clayjohn/dev/godot/./core/variant/variant.h:855)
[32] DisplayServerX11::_dispatch_input_event(Ref<InputEvent> const&) (/home/clayjohn/dev/godot/platform/linuxbsd/x11/display_server_x11.cpp:4012)
[33] DisplayServerX11::_dispatch_input_events(Ref<InputEvent> const&) (/home/clayjohn/dev/godot/platform/linuxbsd/x11/display_server_x11.cpp:3989)
[34] Input::_parse_input_event_impl(Ref<InputEvent> const&, bool) (/home/clayjohn/dev/godot/core/input/input.cpp:772)
[35] Input::flush_buffered_events() (/home/clayjohn/dev/godot/core/input/input.cpp:1044)
[36] DisplayServerX11::process_events() (/home/clayjohn/dev/godot/platform/linuxbsd/x11/display_server_x11.cpp:5103)
[37] OS_LinuxBSD::run() (/home/clayjohn/dev/godot/platform/linuxbsd/os_linuxbsd.cpp:958)
[38] /home/clayjohn/dev/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm(main+0x18e) [0x5863ab6bd93e] (/home/clayjohn/dev/godot/platform/linuxbsd/godot_linuxbsd.cpp:86)
[39] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x731fd6e29d90] (??:0)
[40] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x731fd6e29e40] (??:0)
[41] /home/clayjohn/dev/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm(_start+0x25) [0x5863ab6bd6e5] (??:?)
-- END OF BACKTRACE --
================================================================
```

**Review by clayjohn — COMMENTED** [review]
I don't know why, but when baking with this PR in the Forward+ backend the resulting lightmap is completely black. When baking the MRP, with 4.3 dev 5 I get the same result that [Calinou did](https://github.com/godotengine/godot/issues/77590#issuecomment-1939755066), but with this PR I get a totally black lightmap. 

Have you done something to the MRP to make it work with this PR?

File: modules/lightmapper_rd/lightmapper_rd.cpp (hunk 2 above)
Comment (guerro323) [reply modules/lightmapper_rd/lightmapper_rd.cpp:467]: Ok so I noticed that VSCode didn't added most of my modifications to git for the Compatibility backend, I'll fix that tomorrow

**Comment by guerro323** [comment]
> Have you done something to the MRP to make it work with this PR?
No, except for the second screenshot where I needed to tweak the point lights size.

**Comment by Calinou** [comment]
> but with this PR I get a totally black lightmap.

This might be due to https://github.com/godotengine/godot/issues/90627.

**Comment by clayjohn** [comment]
> > but with this PR I get a totally black lightmap.
> 
> This might be due to #90627.

Just tested, you are right. Fixing https://github.com/godotengine/godot/issues/90627 makes this PR work perfectly in the Forward+ renderer!

File: servers/rendering/renderer_rd/storage_rd/particles_storage.h [hunk 3]
```diff
@@ -360,6 +360,7 @@ class ParticlesStorage : public RendererParticlesStorage {
 		virtual void set_code(const String &p_Code);
 		virtual bool is_animated() const;
 		virtual bool casts_shadows() const;
+		virtual RendererRD::MaterialStorage::ShaderData::CullMode get_cull_mode() const;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/particles_storage.h:363]: We shouldn't expose this, instead we should read the cull mode from the shader directly.

**Review by Mickeon — APPROVED** [review]
Documentation is fine. It's gonna be really cool to implement this in due time.

**Comment by Calinou** [comment]
@guerro323 Are you available to rebase this PR to fix merge conflicts? If not, let me know and I can label this as `salvageable` :slightly_smiling_face:

**Comment by guerro323** [comment]
> @guerro323 Are you available to rebase this PR to fix merge conflicts? If not, no worries – let me know and I can label this as `salvageable` 🙂

Yes, I've just done an oopsie when trying to use the git's stash function to see if the missing code for compatibility was here (spoiler: it wasn't) and broke my branch.

I'll get it done for this week along with some improvements I found while making my own raytracer.

**Comment by jcostello** [comment]
@guerro323 understandable. If its not much to ask, can you rebase the branch?
